### ISSUE TITLE
wrong output with Arabic language

### ISSUE THREAD

**Issue by idrisalshikh** [githubharald/CTCWordBeamSearch#9]
I'm using modified  Keras OCR example  the prediction code is [here](https://colab.research.google.com/drive/1iThzCxnGELv9KFBttUjRfkSZYwieq25z)

also how I save the .csv file 
when I tray to apply your code main.py I get wrong output I'm relay not knowing what is the problem this is the data I used to test the code
[test.zip](https://github.com/githubharald/CTCWordBeamSearch/files/2473963/test.zip)
notes when I open the text files in your data [bentham] and save the files as utf8 the output slightly change 
Decoding 3 samples now.
Sample: 1
Filenames: ../data/bentham/mat_0.csv|../data/bentham/gt_0.txt
Result:       "an"
Ground Truth: "﻿brain."
Editdistance: 5
Accumulated CER and WER so far: CER: 0.7142857142857143 WER: 1.0
Sample: 2
Filenames: ../data/bentham/mat_1.csv|../data/bentham/gt_1.txt
Result:       "supposed"
Ground Truth: "﻿supposed"
Editdistance: 1
Accumulated CER and WER so far: CER: 0.375 WER: 1.0
Sample: 3
Filenames: ../data/bentham/mat_2.csv|../data/bentham/gt_2.txt
Result:       "submitt,both,mental,and.﻿brain"﻿brain"
Ground Truth: "﻿submitt, both mental and corporeal, is far beyond
Editdistance: 32
Accumulated CER and WER so far: CER: 0.5066666666666667 WER: 0.75

**Comment by githubharald** [comment]
1. remove whitespace character from wordChars.txt, as whitespace does not occur inside words but just in between
2. are all words of the ground-truth text also contained inside the corpus.txt? 
3. you can first delete all entries in the data directory (e.g. bentham) and then insert your data, then the program only reads and decodes the one sample you are testing
4. I'm using UTF8 without BOM for Bentham - check that your files do the same

**Comment by idrisalshikh** [comment]
I removed the whitespace 
yes all word in the corpus
I'm using the test data I remove your data and replace it with mine
I also map the latter to English to avoid any problem caused by the encode 
the files attached 
[test.zip](https://github.com/githubharald/CTCWordBeamSearch/files/2475388/test.zip)
can you test the file in your system 
note keras model require include  the whitespace in alphabet

**Comment by githubharald** [comment]
Just to be sure - the CTC blank label is the last entry (as expected by word beam search) in your neural network output? 

I looked through the files:
1. Remove BOM, as this adds some extra bytes at the beginning of the file causing garbage
2. The whitespace must be occur in the chars.txt, but not in the wordChars.txt (the algorithm has to know which characters appear in the neural network output [chars.txt], and it also has to know which of these characters can form words [wordChars.txt]).
3. I suggest setting useNGrams=False in main.py, because there is not enough text in the corpus.txt to learn the bigram probabilities

Result with the changes mentioned above, I also tried vanilla (standard) beam search and best path decoding from the [CTCDecoder repository](https://github.com/githubharald/CTCDecoder):
```
Ground Truth:        "qbm allh gly"
Word beam search:    " qbm allh gly alrqmn alrqbm alr"
Vanilla beam search: "aeqbm alh gly a eaeaeaeacbchc"
Best path decoding:  "aqbm allh gly "
```

So there seems some information "hidden" at the end of the neural network output (after the word "gly"), which is discovered by more advanced algorithms like beam search, but not best path decoding.

**Comment by idrisalshikh** [comment]
thank you for your help , I'm interested in beam search with language model I have small domain so I need the output to be restricted to the language model I set the smoothing to false, what I get now is like this 
the model output is three word {word1,word2,word3} as you mention with best path it get the output correctly also there is garbage text after the three word but this text is not in the language model the code  now add forth word4 from the bi-gram while this word not in the model output the extra word now is the problem how I can modified the code to restrict the output to the recognize word from the mode which near to the langue model

**Comment by githubharald** [comment]
as I said: if best path decoding (which is an approximation) gives better results than beam search decoding, then something with the neural network output is strange. 

Further, you have to test the algorithm on a large set of data (let's say at least 100 or more samples) and then see if best path decoding still performs better.

**Comment by idrisalshikh** [comment]
I test the code with 100 samples 


the sum for edit distance for best path for all samples =83
beam search with   useNGrams=True   take around 15 second for each sample and the output is white space 
beam search with   useNGrams=False   edit distance=499
best path edit distance =83

using keras backend  K.ctc_decode from  [keras ocr example](https://github.com/keras-team/keras/blob/master/examples/image_ocr.py) get the flowing results
when greedy= true which I think it is equal to best path edit distance = 1486
when greedy=false which I think it is equal to beam search edit distance =81 
I attached the files(ground truth ,nn output, corpu) you can take a look 
thank you for your help 
[file.zip](https://github.com/githubharald/CTCWordBeamSearch/files/2488331/file.zip)

**Comment by githubharald** [comment]
sorry but I really don't have the time to debug this.
